cip/CIP2016-06-14-Harmonize-comparability-orderability-equality-and-equivalence.adoc

```diff
@@ -0,0 +1,233 @@
+* Lists containing `null` values are incomparable, i.e. comparing them with any other value yields `null`.
+* Maps are compared as if they where lists of lists, where the inner lists are the key-value pairs of the maps, ordered in ascending sort order for the keys. Comparing a map to a value that is not a map always yields `null`.
+* Maps containing explicit `null` value entries are incomparable, i.e. comparing them with any other value yields `null`.
+* The comparison order for nodes (and relationships resp.) is undefined, and an implementation may choose an order for these freely, e.g. by using some kind of id. This order must not change during the execution of the query (i.e. it must be stable in the presence of updates). Comparing a node to a relationship always yields `null`.
+* Paths are compared as if they where a list of alternating nodes and relationships of the path from the start node to the end node. Comparing a path to a value that is not a path always yields `null`.
+* Implementations may choose to define suitable comparability rules for values of additional, non-canonical types. However comparing a value of an additional, non-canonical type to any of the canonical types always yields `null`.
+* Comparing any kind of value not covered above to an arbitrary other value, always yields `null`. This includes comparing a non-`null` value to a `null` value.
+
+=== Orderability
+
+We further propose that orderability be defined between any pair of values, where the result is always `true` or `false`; i.e. always defined.
+To accomplish this, there must be a pre-determined type order and each value must fall under exactly one type in this order. We propose to use the following ascending global sort order of types:
+
+* `NODE`
+* `RELATIONSHIP`
+* `PATH`
+* `MAP`
+* `LIST OF ANY?`
+* `STRING`
+* `BOOLEAN`
+* `NUMBER` (`NaN` is treated as the largest number in orderability only, i.e. it is put after positive infinity)
+* `VOID` (i.e. the type of `null`)
+
+Within the types, orderability defers to comparability with two exceptions:
+* Two `null` values are treated as the same value in accordance with the global sort order
+* Two `NaN` values are treated as the same value in accordance with the global sort order
+
+The accompanying descending global sort oder is the same order in reverse (i.e. it runs from `VOID` to `NODE`).
+
+Any additionally introduced non-canonical types introduced by an implementation should be inserted between `PATH` and `MAP` in the global sort order. As an exception, additionally introduced non-canonical number types may be integrated under the `NUMBER` type in the global sort order.
+
+=== Equality ===
+
+We propose that equality and comparability must be aligned, i.e. `l = r` if and only if `l <= r && l >= r`.
+
+To achieve this, we unify the difference between equality and equivalence, by making lists containing `null` values follow the same difference as the `null` values themselves.
```

Comment:
"To achieve this, we unify the difference between equality and equivalence, by making" 
--->
"We thus unify the difference between equality and equivalence by making"